%% UNCOMMENT FOR SLIDES
\input{sections/preamble_slides}
%% UNCOMMENT FOR HANDOUTS
%\input{sections/preamble_handouts}
%% GENERIC STYLE SETTINGS BELOW
\input{sections/preamble_style}
% LISTINGS SETTING
\input{sections/preamble_codelistings}

%%%
% TITLE PREAMBLE
\title[Intro to Bioinformatics] % (optional, only for long titles)
{An Introduction to Bioinformatics Tools}
\subtitle{Part 3: Workshop}
\author[Pritchard, Cock] % (optional, for multiple authors)
{Leighton~Pritchard \and Peter~Cock}
\institute[The James Hutton Institute] % (optional)
{
  Information and Computational Sciences\\
  The James Hutton Institute
}
\date[May 2014] % (optional)
{Bioinformatics Training, 29$^{th}$, 30$^{th}$ May 2014}
\subject{Bioinformatics}

%%%
% TOC
\input{sections/preamble_toc}

%%%
% START DOCUMENT
\begin{document}

\frame[plain]{\titlepage}
 
%%%
% SECTION: Introduction
\section{Introduction}
\subsection{Introduction}
\begin{frame}
  \frametitle{Learning Outcomes}
  \begin{itemize}
    \item Workshop example: bacterial genome annotation \\
             (because they're small and data easy to handle)
    \begin{itemize}
      \item The role of biological insight in a bioinformatics workflow
      \item Visual interaction with sequence data
      \item Using alternative tools
      \item Comparison of tools and outputs
      \item Online tools for automated function prediction
    \end{itemize}
  \end{itemize}
\end{frame} 
  
% Workshop Outline
\begin{frame}
  \frametitle{What You Will Be Doing}
  Illustrative example of concepts: Functional annotation of a draft bacterial genome
  \begin{enumerate}
    \item Gene prediction
    \item Genome comparisons
%% Skipping section for time
%    \item Gene functional annotation
    \item Gene comparisons
  \end{enumerate}
\end{frame}

%%%
% SECTION: Loading workshop data into Artemis
\section{Workshop Data}
  \include{sections/subsection_locateworkshopdata}
  \include{sections/subsection_loadworkshopdata}

%%%
% SECTION: Gene Prediction
\section{Gene Prediction}
  \include{sections/subsection_genepredictionintro}
  %% Replaced with basic gene finding
  %\include{sections/subsection_orfprediction}
  \include{sections/subsection_basicgenefinding}
  \include{sections/subsection_prokaryoticgeneprediction}
  \include{sections/subsection_assessinggeneprediction}

%%%
% Whole genome comparisons
\section{Genome Comparisons}
  \include{sections/subsection_blastgenomecomparison}
  \include{sections/subsection_mummergenomecomparison}

% Whole genome comparison conclusions
\subsection{Conclusion}
\begin{frame}
  \frametitle{Genome Alignments}   
  \framesubtitle{Lessons learned}   
  \begin{itemize}
    \item Alignment result depends on algorithm, and parameter choices
    \item Some algorithms/parameter sets more sensitive than others
    \item Appropriate visualisation is essential
  \end{itemize}
  Much more detail at \url{http://www.slideshare.net/leightonp/comparative-genomics-and-visualisation-part-1}
\end{frame}

%%%%
%% SECTION: Functional Prediction
% Commented out for time
%\section{Functional Prediction}
%  \include{sections/subsection_functionalprediction}

%%%
% SECTION: Gene Comparisons
\section{Gene Comparisons}
  \include{sections/subsection_reciprocalblast}
  %% Commented out for time
  % \include{sections/subsection_hmmer}
  %% Commented out for time
  % \include{sections/subsection_msa}


%%%
% SECTION: Conclusion
\section{Conclusions}
  \subsection{Conclusions}
\begin{frame}[fragile]
  \frametitle{In Conclusion}
  \begin{itemize}
    \item The tools you will need to use will be task-dependent, but some things are universal$\ldots$
    \begin{itemize}
      \item Good experimental design (including BLAST searches, etc.)
      \item Keeping accurate records for reproduction/replication
      \item Validation/sanity checking of results
      \item Comparison and benchmarking of methods
      \item (Cross-)validation of predictive methods
    \end{itemize}
     Remember: everything gets easier with practice, so practice lots!   
  \end{itemize}
\end{frame}


% etc
\end{document}
