% Exercise for Second Golden Rule of Bioinformatics
%
% The aim of this exercise is to demonstrate that even a well-known,
% highly trusted bioinformatics tool, such as BLAST, does not do the 
% biology for you. In particular it does not check your input or database
% sequences for validity, or that you chose an appropriate set of
% sequences to answer your questions correctly.
% It also illustrates how easy it is to tell a biologically-convincing story
% from rubbish.

\subsection{Golden Rule 2 Exercise}
\begin{frame}
  \frametitle{Exercise 2}
  \framesubtitle{Instructions}    
  \begin{itemize}
    \item You are in group A, B, C or D - this decides your database\\
    \texttt{dbA}, \texttt{dbB}, \texttt{dbC}, \texttt{dbD}
    \item You will use \texttt{BLAST} at the command-line to analyse your data
  \end{itemize}
\end{frame}

% [fragile] frames must end with \end{frame} directly following a newline, or they break!
\begin{frame}[fragile]
  \frametitle{Exercise 2}
  \framesubtitle{Running \texttt{BLAST}}
  \begin{itemize}
    \item Change directory to the \texttt{ex2\_blast} directory
    \item Run \texttt{BLAST} with the appropriate database
  \end{itemize}
\begin{lstlisting}[language=bash]
$ cd ../ex2_blast
$ blastp -num_alignments 1 -num_descriptions 1 -query query.fasta -db dbA
\end{lstlisting}    
\end{frame}

% [fragile] frames must end with \end{frame} directly following a newline, or they break!
\begin{frame}[fragile]
  \frametitle{Exercise 2}
  \framesubtitle{\texttt{BLAST} output}
  \begin{tiny}
  \begin{verbatim}
Query= query protein sequence

Length=400
                                                                      Score
Sequences producing significant alignments:                          (Bits)

  PITG_08491T0 Phytophthora infestans T30-4 choline transporter-l...  34.3


> PITG_08491T0 Phytophthora infestans T30-4 choline transporter-like 
protein (441 aa)
Length=486

 Score = 34.3 bits (77), Method: Compositional matrix adjust.
 Identities = 22/69 (32%), Positives = 38/69 (55%), Gaps = 4/69 (6%)

Query  106  EVILPMMYQFALKPSFADVINDYKPYSKHTAGVSDQELKGEATTWMLADKNSRMKAFLSQ  165
            E+++PM+Y   L   F   ++ Y P   HTA ++  EL+G   T ++A+  S +  F ++
Sbjct  40   ELMVPMLYSLYLVVLFHLPVSAYYP---HTASMTAHELQGAVITILVAETPSIIIQF-AK  95

Query  166  IKTKSNSSE  174
              T SN S+
Sbjct  96   CHTSSNISQ  104
  \end{verbatim}   
  \end{tiny} 
\end{frame}

\begin{frame}
  \frametitle{Exercise 2}
  \framesubtitle{Results}
  \begin{itemize}
    \item What is a reasonable E-value threshold to call a 'match'?
    \begin{itemize}
      \item 1e-05, 0.001, 0.1, 10?
    \end{itemize}
  \end{itemize}
  \begin{center}
  \begin{tabular}{r|l|l|l|l}
            & dbA & dbB & dbC & dbD \\
	  \hline
	  \hline
	  E-value &      &  &  & \\
  \end{tabular}
  \end{center}
\end{frame}

\begin{frame}
  \frametitle{Exercise 2}
  \framesubtitle{Results}
  \begin{itemize}
    \item What is a reasonable E-value threshold to call a 'match'?
    \begin{itemize}
      \item 1e-05, 0.001, 0.1, 10?
    \end{itemize}
  \end{itemize}
  \begin{center}
  \begin{tabular}{r|l|l|l|l}
	   & dbA & dbB & dbC & dbD \\
	  \hline
	  \hline
	  E-value & 0.45 & 0.002 & 4e-06 & 0.019 \\
  \end{tabular}
  \end{center}
  \begin{itemize}
    \item Five orders of magnitude difference in E-value, depending on database choice - Why?
  \end{itemize}    
\end{frame}

\begin{frame}
  \frametitle{Exercise 2}
  \framesubtitle{Results}
  \begin{itemize}
    \item E-values depend on database size
    \item Bit score and alignment do not depend on database size
  \end{itemize}
  \begin{center}
  \begin{tabular}{r|l|l|l|l}
	   & dbA & dbB & dbC & dbD \\
	  \hline
	  \hline
	  E-value & 0.45 & 0.002 & 4e-06 & 0.019 \\
	  Bit score & 34.3 & 34.3 & 34.3 & 34.3 \\
	  \hline
	  Sequences & 100,001 & 501 & 1 & 5,001 \\
	  Letters & 48,650,486 & 210,866 & 486 & 2,066,510 	  
  \end{tabular}
  \end{center}
\end{frame}

\begin{frame}
  \frametitle{Exercise 2}
  \framesubtitle{Results}    
  \begin{itemize}
    \item<1-> E-values differ, but the query matches a choline transporter-like 
protein quite well$\ldots$
    \item<2-> \emph{Doesn't it?}
    \item<1-> After all, a biological match is a biological match$\ldots$
    \item<2-> \emph{Isn't it?}
  \end{itemize}
\end{frame}

% [fragile] frames must end with \end{frame} directly following a newline, or they break!
\begin{frame}[fragile]
  \frametitle{Exercise 2}
  \framesubtitle{\texttt{BLAST} output}
  \begin{tiny}
  \begin{verbatim}
Query= query protein sequence

Length=400
                                                                      Score     E
Sequences producing significant alignments:                          (Bits)  Value

  PITG_08491T0 Phytophthora infestans T30-4 choline transporter-l...  34.3    4e-06


> PITG_08491T0 Phytophthora infestans T30-4 choline transporter-like 
protein (441 aa)
Length=486

 Score = 34.3 bits (77),  Expect = 4e-06, Method: Compositional matrix adjust.
 Identities = 22/69 (32%), Positives = 38/69 (55%), Gaps = 4/69 (6%)

Query  106  EVILPMMYQFALKPSFADVINDYKPYSKHTAGVSDQELKGEATTWMLADKNSRMKAFLSQ  165
            E+++PM+Y   L   F   ++ Y P   HTA ++  EL+G   T ++A+  S +  F ++
Sbjct  40   ELMVPMLYSLYLVVLFHLPVSAYYP---HTASMTAHELQGAVITILVAETPSIIIQF-AK  95

Query  166  IKTKSNSSE  174
              T SN S+
Sbjct  96   CHTSSNISQ  104
  \end{verbatim}   
  \end{tiny} 
\end{frame}

\begin{frame}
  \frametitle{Exercise 2}
  \framesubtitle{Remember Golden Rule 1?}    
  \begin{itemize}
    \item<1-> Sequence accessions (\texttt{PITG\_?????T0}) \emph{are} correct in the databases
    \item<2-> Sequence functional descriptions are randomly shuffled: lengths do not match in \texttt{BLAST} output
    \item<3-> \texttt{dbA} contains only three different sequences: two are repeated 50,000 times
    \item<4-> \texttt{query.fasta} is random sequence, not a real protein
    \begin{itemize}
      \item<4-> Shuffled from all \textit{P. infestans} proteins
      \item<4-> No \texttt{nr} or \texttt{PFam} matches        
    \end{itemize}
  \end{itemize}
\end{frame}
